drivers/timer/intel_adsp_timer.c

```diff
@@ -58,18 +56,26 @@ const int32_t z_sys_timer_irq_for_test = TIMER_IRQ; /* See tests/kernel/context
 
 static void set_compare(uint64_t time)
 {
-	/* Disarm the comparator to prevent spurious triggers */
-	*WCTCS &= ~DSP_WCT_CS_TA(COMPARATOR_IDX);
+	const struct device *dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, syscon));
+	uint32_t wctcs;
 
-	*COMPARE_LO = (uint32_t)time;
-	*COMPARE_HI = (uint32_t)(time >> 32);
+	/* Disarm the comparator to prevent spurious triggers */
+	syscon_read_reg(dev, ADSP_DSPWCTCS_OFFSET, &wctcs);
+	wctcs &= ~DSP_WCT_CS_TA(COMPARATOR_IDX);
+	syscon_write_reg(dev, ADSP_DSPWCTCS_OFFSET, wctcs);
+	syscon_write_reg(dev, ADSP_DSPWCT0C_OFFSET, (uint32_t)time);
+	syscon_write_reg(dev, ADSP_DSPWCT0C_OFFSET + 4, (uint32_t)(time >> 32));
 
 	/* Arm the timer */
-	*WCTCS |= DSP_WCT_CS_TA(COMPARATOR_IDX);
+	syscon_read_reg(dev, ADSP_DSPWCTCS_OFFSET, &wctcs);
+	wctcs |= DSP_WCT_CS_TA(COMPARATOR_IDX);
+	syscon_write_reg(dev, ADSP_DSPWCTCS_OFFSET, wctcs);
```

Comment:
Wouldn't be better to cache these registers in the iniatialization to avoid these calls ? Even from kernel space we still have some overhead ... and to be honest this is not looking more readable imho.

Reply:
what overhead are talking about here?

regarding readability, it is a matter of taste IMO. In the future this will also help with MMU. In this particular case we can probably optimize and use the non-syscall variant of the io functions for reading/writing registers.

Reply:
Mostly agree with @ceolin.  This version is going to generate a TON more code, including four function pointer invocations with null checks (indirected calls happen to be particularly expensive on xtensa).  Whether it will be measurably impactful performance-wise is more of a question; we call the timer system a lot when IPC state changes, but that doesn't happen much in SOF which has a fixed timer architecture.

Reply:
right, the mmu part of syscon I completely missed and raised yesterday in the arch. meeting with @carlocaione. Syscon was recommended as a solution to how intel adsp does things, i.e. the shim registers which is not specific to one device an used across all drivers and soc layers. Trying to be a good DTS citizen, but I might do this using some other way and that is lower level than syscon.

Reply:
agree that syscon is introducing some overhead here. A couple of ideas here: (1) you can call `syscon_get_base()`, cache the address in your driver and use that in the critical paths or (2) we can extend the API with now low-overhead calls

Reply:
keeping the syscon reference in DTS but not using the syscon driver for now. Maybe we need a `syscon_light` :)

Reply:
ack for a ```syscon_light()```. The big benefit of syscon and the likes vs structures is that 
1) Can be traced. Its often critical to see HW IO ordering and timestamped.
2) Can be easily grepped. Easy to visualize who is doing what and where wrt HW IO. (structures can be named differently in the codebase making grepping difficult)
3) Can be SPARSE checked. i.e are clients passing the correct ptr type (important on Intel HW with cache architecture).

Fwiw, the SOF equivalent is a light static inline with no NULL check.